### PR TITLE
Include the option for imaginary parts in the UCC ansatz 

### PR DESCRIPTION
**Pull request by tanvipg**
### Summary
A feature added to the ucc.py file to include imaginary contributions to the ucc wave function along with the real parts. To allow for complex values in the Hamiltonian, changes to the 'so_matrix' initiation in the 'two_body_electronic_integrals.py' were added as well. 

### Details and comments
1. Added dtype = 'complex' in so_matrix creation in 'two_body_electronic_integrals.py'
2. Added a boolean flag, 'include_imaginary', in the 'ucc.py' file initialization to include the option of creating the imaginary part of the ucc wavefunction with twice the number of free parameters as described below

Screen Shot 2021-12-07 at 6 29 49 PM
3. Modified the '_build_fermionic_excitation_ops()' in 'ucc.py' file to incorporate the additions described above

### CODE REVIEW

File: qiskit_nature/circuit/library/ansatzes/ucc.py
Comment (tanvipg) [inline qiskit_nature/circuit/library/ansatzes/ucc.py:432]: Suggestion: `op_plus = -1*(op+op_adj)`

While the -1 will get incorporated in parameter optimization, the negative sign is consistent with `PauliTrotterEvolution` class

Comment (pbark) [reply qiskit_nature/circuit/library/ansatzes/ucc.py]: fixed in 21a3456

**Comment by a-matsuo** [comment]
I think unittests are required for this new feature since the feature is totally new.

File: qiskit_nature/circuit/library/ansatzes/ucc.py
Comment (pbark) [inline qiskit_nature/circuit/library/ansatzes/ucc.py:171]:
```suggestion
```

**Review by pbark — CHANGES_REQUESTED** [review]

File: qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py
Comment (woodsp-ibm) [inline qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py:167]: This change to complex will end up consuming a bunch more memory (double) since commonly we have real values for these integrals I think. I am also trying to figure out how this change factors in since the UCC operators are built directly as FermionicOps and not via integrals right.

File: qiskit_nature/circuit/library/ansatzes/ucc.py
Comment (mrossinek) [inline qiskit_nature/circuit/library/ansatzes/ucc.py:177]:
```suggestion
            include_imaginary: boolean flag which when set to 'True' expands the ansatz to include
                imaginary parts using twice the number of free parameters.
```
This part is not relevant since this is non-public API.

I am also wondering whether we could call this, `include_complex_conjugate` which is a better description, imo

File: qiskit_nature/circuit/library/ansatzes/uccsd.py
Comment (mrossinek) [inline qiskit_nature/circuit/library/ansatzes/uccsd.py:38]: The same setting should also be exposed in the other subclasses of `UCC`

File: qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py
Comment (mrossinek) [inline qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py:167]: I agree with @woodsp-ibm's original comment, that this is duplicating the memory requirements unnecessarily in most cases.
Thus, this should be a configurable setting.

We can either:
1. add a setting in `qiskit_nature.settings`
2. add an `__init__` argument to these integrals (since the `OneBody` case should probably deal with this, too) which would then also need to be propagated down to the `ElectronicEnergy`

I think, I would prefer the former option.

@pbark any thoughts on this?

File: test/circuit/library/ansatzes/test_ucc.py
Comment (mrossinek) [inline test/circuit/library/ansatzes/test_ucc.py:29]: Is the code that was commented here not needed anymore? If yes, the comment ought to be removed

File: test/circuit/library/ansatzes/test_ucc.py
Comment (mrossinek) [inline test/circuit/library/ansatzes/test_ucc.py:220]: I don't think these tests were planned to be removed. Looks like a bad merge since these were only added back in March

**Review by mrossinek — CHANGES_REQUESTED** [review]

File: releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml:6]: @mrossinek @declanmillar So setting imaginary here doubles up the number of parameters. How does this impact/affect the recent InitialPoint work that would set the starting point value for these?

File: qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py
Comment (mrossinek) [reply qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py:167]: Thanks for the reply. Is it needed at all for this PR to work as intended though? If not, I would like to leave this change out and consider this scenario as part of #665 instead

File: releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml
Comment (declanmillar) [reply releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml:6]: @mrossinek, @woodsp-ibm, @tanvipg, I've opened a PR [here](https://github.com/Qiskit/qiskit-nature/pull/455) that should patch this behavior. Let me know if you spot any issues with it.

Comment (mrossinek) [reply releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml:6]: Thanks a lot, @declanmillar! The patch looks good, thanks for also including a unittest!
I pinged Tanvi on Slack, too, in order to hopefully get your PR merged asap :+1:

File: qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py
Comment (mrossinek) [reply qiskit_nature/properties/second_quantization/electronic/integrals/two_body_electronic_integrals.py:167]:
```suggestion
        so_matrix = np.zeros([2 * s for s in self._matrices[0].shape])
```
After another offline check with Panos, we decided to keep this change out of this PR and instead introduce this as a configurable option as part of the operator-overhaul (#665)

**Review by mrossinek — APPROVED** [review]
This LGTM now (assuming CI passes). I'll await another review by @pbark or @woodsp-ibm before merging :+1:

File: releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml
Comment (woodsp-ibm) [inline releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml:6]: I think we can make things clearer here following the first sentence. While we know parameter here means parameter of the UCC ansatz circuit it could be talking about a method parameter here as it seems a bit vague to me. How about the second sentence being replaced by something like the following which also spells out what happens when you set it  i.e. doubles the number of parameters. as that can have implications for optimization.

`When True an extra ansatz parameter is added to each excitation that controls the imaginary contribution to its evolution. This setting then doubles the total number of ansatz parameters, as compared to the default setting of False.`

Comment (mrossinek) [reply releasenotes/notes/ucc_imaginary-8d7d9c3322899a34.yaml]: Thanks, I updated the release note with your suggestion :+1: 

With regards to the example, this would be something for @pbark or @tanvipg to answer

**Comment by mrossinek** [comment]
With complex integrals no longer being an issue in the stack (see also #1144), I have updated this PR to the latest changes.
I am not confident that the proposed inclusion of this feature in the SUCCD case will work as intended since this was applied to the code version prior to the addition of the `mirror` feature in #729. Thus, I refrained from exposing this keyword argument in that specific subclass in this PR.

**Review by woodsp-ibm — APPROVED** [review]

**Pull request merged**
